pkg/scheduler/service/cleaner.go

```diff
@@ -34,14 +45,67 @@ func (c *cleaner) Run(ctx context.Context, transition *ClusterStatusTransition,
 		case <-ticker.C:
 			c.purgeReconciliations(transition, config)
 		case <-ctx.Done():
-			c.logger.Info("Stopping cleaner because parent context got closed")
+			c.logger.Info("[CLEANER] Stopping because parent context got closed")
 			ticker.Stop()
 			return nil
 		}
 	}
 }
 
 func (c *cleaner) purgeReconciliations(transition *ClusterStatusTransition, config *CleanerConfig) {
+
+	c.logger.Info("[CLEANER] Process started")
+
+	if config.KeepLatestEntitiesCount > 0 {
+		c.purgeReconciliationsNew(transition, config)
+	} else {
+		c.purgeReconciliationsOld(transition, config)
+	}
+
+	c.logger.Info("[CLEANER] Process finished")
+}
+
+//Purges reconciliations using rules from: https://github.com/kyma-incubator/reconciler/issues/668
+func (c *cleaner) purgeReconciliationsNew(transition *ClusterStatusTransition, config *CleanerConfig) {
+	now := time.Now()
+
+	latestReconciliations, err := c.getLatestReconciliations(config.keepLatestEntitiesCount(), transition)
+	if err != nil {
+		c.logger.Errorf("Cleaner failed to get last %d reconciliations: %s", config.keepLatestEntitiesCount(), err.Error())
+	}
+
+	if len(latestReconciliations) < config.keepLatestEntitiesCount() {
+		//Nothing to clean up
+		return
+	}
+
+	oldestInRange := findOldestReconciliation(latestReconciliations)
```

Comment:
I think the returned `latestReconciliations` array is already sorted by `{"Created": "DESC"}`, you can just use the last one as oldest

Reply:
I am aware of that, this is just to be sure in case someone in the future breaks the contract and the returned list is not sorted as we expected. I think spending few dozens microseconds on iteration is worth it. After all the oldest record found is then taken as the reference for removal of really important entities (auditing, on-call debugging etc). In cases like this I would say a "double-check" is justified.